### ISSUE TITLE
TypeError at /login/finalize/ authenticate() takes exactly 3 arguments (2 given)

### ISSUE THREAD

**Issue by cdiconzaVL** [discolabs/django-shopify-auth#34]
![typeerror](https://cloud.githubusercontent.com/assets/28960188/26462912/a8689b92-4150-11e7-9c76-e7d9f0d415dc.png)

Hi there, I followed your video guide and also tried downloading a zip of the example auth-app project but I get this error in both cases. 

After navigating to https://.../accounts/login/?next=/ and clicking install, it tells me that authenticate failed in backends.py because only 2 arguments were given.  
![installed](https://cloud.githubusercontent.com/assets/28960188/26463074/25be1b76-4151-11e7-9e1b-27ca23fc09f7.png)
It definitely installed but my user is not being authenticated, {{user}} remains as "anonymousUser".

What might cause this?

I put some print lines in your site-packages/shopify_auth/views.py file. It looks like shopify_session.token is None in the views.py file, and args & kwargs are empty as well. However, in backends.py I can see that the token definitely has a value. 

Any help would be greatly appreciated. Thank you.

I copied the traceback:

-----------------------------------------------------------------------------------------------------------------
Environment:

Request Method: GET
Request URL: https://2ea3be8d.ngrok.io/login/finalize/?code=5407027b4d609d4bcdd6729784e2dd4e&hmac=397a57b2ba1953c65ec88e26b0f9adca58eac8cfe611c61efb1fc444cdb0a42a&shop=vlpythonteststore.myshopify.com&timestamp=1495733623

Django Version: 1.11.1
Python Version: 2.7.13
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'auth_app',
 'shopify_auth']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware']

Traceback:

File "C:\Python27\lib\site-packages\django\core\handlers\exception.py" in inner
  41.             response = get_response(request)

File "C:\Python27\lib\site-packages\django\core\handlers\base.py" in _get_response
  187.                 response = self.process_exception_by_middleware(e, request)

File "C:\Python27\lib\site-packages\django\core\handlers\base.py" in _get_response
  185.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "C:\Python27\lib\site-packages\django\contrib\auth\decorators.py" in _wrapped_view
  23.                 return view_func(request, *args, **kwargs)

File "C:\Python27\lib\site-packages\shopify_auth\views.py" in finalize
  83.     user = auth.authenticate(myshopify_domain=shopify_session.url, token=shopify_session.token)

File "C:\Python27\lib\site-packages\django\contrib\auth\__init__.py" in authenticate
  101.             user = backend.authenticate(*args, **credentials)

File "C:\Python27\lib\site-packages\shopify_auth\backends.py" in authenticate
  11.         user = super(ShopUserBackend, self).authenticate(remote_user=myshopify_domain)

Exception Type: TypeError at /login/finalize/
Exception Value: authenticate() takes exactly 3 arguments (2 given)

**Comment by gavinballard** [comment]
Hi @cdiconzaVL, can you confirm if this occurs if your Django version is 1.9? That's the last version that I've used / tested the package with.

To be honest I don't have a lot of time to maintain this package as we don't actively use it at Disco any more, although if you do identify an issue I'm always happy to merge tested PRs.

**Comment by cdiconzaVL** [comment]
Thanks for the response. I installed DJango 1.9, and the authenticate error is not showing up anymore. However, after signing in I just get redirected back to the login page. I think my user is still an AnonymousUser, using printlines it is showing up as None in views.py finalize method.
![usernone](https://cloud.githubusercontent.com/assets/28960188/26552317/a245c1de-4454-11e7-88e4-74bd941ab933.png)

**Comment by cfc603** [comment]
In [Django 1.11 the request argument was added to ModelBackend](https://docs.djangoproject.com/en/1.11/ref/contrib/auth/#django.contrib.auth.backends.ModelBackend.authenticate). Not sure if I'll have time to make a pull request, but the following solves the problem.

```python
# shopify_auth/backends.py
from django.contrib.auth.backends import RemoteUserBackend


class ShopUserBackend(RemoteUserBackend):

    def authenticate(self, myshopify_domain=None, token=None, **kwargs):
        if not myshopify_domain or not token:
            return

        user = super(ShopUserBackend, self).authenticate(
            request=kwargs['request'], remote_user=myshopify_domain
        )
        if not user:
            return

        user.token = token
        user.save(update_fields=['token'])
        return user

```

**Comment by gavinballard** [comment]
@cfc603 Any chance this could be added as a PR?

**Comment by Benoss** [comment]
This should be fixed now, just need to release 0.7
